src/main/java/eu/dissco/orchestration/backend/service/DataMappingService.java

```diff
@@ -0,0 +1,240 @@
+      throw new ProcessingFailedException(e.getMessage(), e);
+    }
+    var dataMapping = buildDataMapping(mappingRequest, 1, userId, handle,
+        Date.from(Instant.now()));
+    repository.createMapping(dataMapping);
+    publishCreateEvent(dataMapping);
+    return wrapSingleResponse(dataMapping, path);
+  }
+
+  private DataMapping buildDataMapping(DataMappingRequest dataMappingRequest, int version,
+      String userId, String handle, Date created) {
+    var id = HANDLE_PROXY + handle;
+    return new DataMapping()
+        .withId(id)
+        .withOdsID(id)
+        .withType("ods:DataMapping")
+        .withOdsType(fdoProperties.getDataMappingType())
+        .withSchemaVersion(version)
+        .withOdsStatus(OdsStatus.ODS_ACTIVE)
+        .withSchemaName(dataMappingRequest.getSchemaName())
+        .withSchemaDescription(dataMappingRequest.getSchemaDescription())
+        .withSchemaDateCreated(created)
+        .withSchemaDateModified(Date.from(Instant.now()))
+        .withSchemaCreator(new Agent().withId(userId).withType(Type.SCHEMA_PERSON))
+        .withOdsDefaultMapping(buildDefaultMapping(dataMappingRequest))
+        .withOdsFieldMapping(buildFieldMapping(dataMappingRequest))
+        .withOdsMappingDataStandard(OdsMappingDataStandard.fromValue(
+            dataMappingRequest.getOdsMappingDataStandard().value()));
+  }
+
+  private void publishCreateEvent(DataMapping dataMapping)
+      throws ProcessingFailedException {
+    try {
+      kafkaPublisherService.publishCreateEvent(mapper.valueToTree(dataMapping));
+    } catch (JsonProcessingException e) {
+      log.error("Unable to publish message to Kafka", e);
+      rollbackMappingCreation(dataMapping);
+      throw new ProcessingFailedException("Failed to create new machine annotation service", e);
+    }
+  }
+
+  private void rollbackMappingCreation(DataMapping dataMapping) {
+    var request = fdoRecordService.buildRollbackCreateRequest(dataMapping.getId());
+    try {
+      handleComponent.rollbackHandleCreation(request);
+    } catch (PidAuthenticationException | PidCreationException e) {
+      log.error(
+          "Unable to rollback handle creation for data mapping. Manually delete the following handle: {}. Cause of error: ",
+          dataMapping.getId(), e);
+    }
+    repository.rollbackMappingCreation(dataMapping.getId());
+  }
+
+  public JsonApiWrapper updateDataMapping(String id, DataMappingRequest dataMappingRequest, String userId,
+      String path)
+      throws NotFoundException, ProcessingFailedException {
+    var currentVersion = repository.getActiveMapping(id);
+    if (currentVersion.isEmpty()) {
+      throw new NotFoundException("Requested data mapping does not exist");
+    }
+    if (isEqual(dataMappingRequest, currentVersion.get())) {
+      return null;
+    } else {
+      var newMapping = buildDataMapping(dataMappingRequest, currentVersion.get().getSchemaVersion() + 1,
```

Comment:
`isEqual()` and `buildDataMapping()` both call the `buildDefault/FieldMapping()` functions. These look a bit more expensive than `buildDataMapping()`. I'd suggest building the dataMapping first, comparing updatedDataMapping with the previous version, and creating the object in the repository if not equal. That way, you can compare the default and field mappings without having to build them twice. 

```
var newMapping = buildDataMapping(dataMappingRequest, currentVersion.get().getSchemaVersion() + 1)
if (isEqual(currentVersion, newMapping)) { // modify to compare field and default mappings directly
return null
}
repository.updateMapping(newMapping)
```

Not sure how big the data mappings get, but it might be nice not to have to build them twice.

Reply:
Yes true, fair point. Changed it so it does the building a bit sooner and then equals.